homeassistant/components/stream/hls.py

```diff
@@ -64,55 +78,184 @@ class HlsPlaylistView(StreamView):
+                        "#EXT-X-PROGRAM-DATE-TIME:"
+                        + segment.start_time.strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3]
+                        + "Z",
+                    ]
+                )
+            # Only include EXT-X-PART data for last 3 segments. The RFC seems
+            # to suggest removing after 3 full segments, but Apple's own
+            # example shows removing after 2 full segments and 1 part one.
+            if ll_hls and i >= 0:
+                for part in segment.parts:
+                    playlist.append(
+                        f"#EXT-X-PART:DURATION={float(part.duration):.3f},URI="
+                        f'"./segment/{segment.sequence}.m4s",BYTERANGE="{len(part.data)}'
+                        f'@{part.http_range_start}"{",INDEPENDENT=YES" if part.independent else ""}'
+                    )
+            if segment.duration:
+                playlist.extend(
+                    [
+                        f"#EXTINF:{float(segment.duration):.3f},",
+                        f"./segment/{segment.sequence}.m4s",
+                    ]
+                )
             last_stream_id = segment.stream_id
+        # Add preload hint
+        # segments is checked above, so segment should be available from the loop iteration
+        # pylint: disable=undefined-loop-variable
+        if ll_hls:
+            if segment.duration:  # Next part belongs to next segment
+                playlist.append(
+                    f'#EXT-X-PRELOAD-HINT:TYPE=PART,URI="./segment/{segment.sequence+1}'
+                    '.m4s",BYTERANGE-START=0'
+                )
+            else:  # Next part is in the same segment
+                start = (
+                    segment.parts[-1].http_range_start + len(segment.parts[-1].data)
+                    if len(segment.parts) > 0
+                    else 0
+                )
+                playlist.append(
+                    f'#EXT-X-PRELOAD-HINT:TYPE=PART,URI="./segment/{segment.sequence}'
+                    f'.m4s",BYTERANGE-START={start}'
+                )
+        return "\n".join(playlist) + "\n"
 
-        return playlist
+    @staticmethod
+    def bad_request(blocking, target_duration):
+        """Return a HTTP Bad Request response."""
+        return web.Response(
+            body=None,
+            status=400,
+            headers={
+                "Cache-Control": f"max-age={(4 if blocking else 1)*target_duration:.0f}"
+            },
+        )
 
-    def render(self, track):
-        """Render M3U8 file."""
-        lines = ["#EXTM3U"] + self.render_preamble(track) + self.render_playlist(track)
-        return "\n".join(lines) + "\n"
+    @staticmethod
+    def not_found(blocking, target_duration):
+        """Return a HTTP Not Found response."""
+        return web.Response(
+            body=None,
+            status=404,
+            headers={
+                "Cache-Control": f"max-age={(4 if blocking else 1)*target_duration:.0f}"
```

Comment:
These Cache-Control headers come from the recommendations in the RFC. From B.1:

>       Blocking Playlist Reload allows longer caching of Playlists
>       without detriment to Clients.  Successful responses to blocking
>       Playlist requests should be cached for six Target Durations.
>       Unsuccessful responses (such as 404s) should be cached for four
>       Target Durations.  Successful responses to non-blocking Playlist
>       requests should be cached for half the Target Duration.
>       Unsuccessful responses to non-blocking Playlist requests should be
>       cached for for one Target Duration.
> 
>       Successful responses to blocking Media segment requests should be
>       cached for six Target Durations.  Unsuccessful responses should be
>       cached for one Target Duration.
> 
>       Origin servers should use Cache-Control headers to communicate the
>       desired cache lifetime.

Reply:
Want to leave a comment about these "magic numbers"?

Reply:
Comment is now on L288